### PR TITLE
Run data flow analysis over field initializers when instrumenting for dynamic analysis

### PR DESCRIPTION
**Pull request by JohnHamby**
When compiling with instrumentation for dynamic analysis, data flow analysis was not being run over initializers incorporated into synthesized constructors. This led to spurious diagnostics claiming that fields with initializers were not being initialized, because the initializations were never seen by flow analysis. This change set corrects that, and adds verification of diagnostics to all instrumentation tests.

@dotnet/roslyn-compiler @jcouv @cston @AlekseyTs @dotnet/testimpact @tannergooding @shyamnamboodiripad @ManishJayaswal

### CODE REVIEW

**Comment by jcouv** [comment]
Thanks for the in-person explanation. This change LGTM.

For other reviewers, the background is that in most cases, field initializer code is injected into constructors only after lowering and so was therefore not getting instrumented. 
With this change, such code is injected before lowering (so it will get instrumented). But that caused a problem with data flow analysis (the initialization code was not analyzed because synthesized methods that return void are apparently skipped), so this change also adds a call to analyze the injected code. 

@gafter @AlekseyTs I wonder whether it would be simpler/cleaner to remove the condition that skips doing data flow analysis on synthesized methods that return void. 
@JohnHamby Could you point us to the method/code where this skip occurs?

**Comment by JohnHamby** [comment]
This change doesn't move the point where initializers are injected into constructors when instrumenting--that happened in an earlier PR. This change just forces flow analysis of the initializers.

FlowAnalysisPass.Rewrite suppresses analysis of synthesized constructors with this code:

            if (method.ReturnsVoid || method.IsIterator || method.IsTaskReturningAsync(compilation))
            {
                // we don't analyze synthesized void methods.
                if ((method.IsImplicitlyDeclared && !method.IsScriptInitializer)

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 1]
```diff
@@ -888,6 +888,14 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
 
                         if (body != null && ((methodSymbol.ContainingType.IsStructType() && !methodSymbol.IsImplicitConstructor) || instrumentForDynamicAnalysis))
                         {
+                            if (instrumentForDynamicAnalysis && methodSymbol.IsImplicitConstructor)
+                            {
+                                // Flow analysis over the initializers is necessary in order to find assignments to fields.
+                                // Bodies of implicit constructors do not get flow analysis later, so the initializers
+                                // are analyzed here.
+                                DataFlowPass.Analyze(_compilation, methodSymbol, analyzedInitializers, diagsForCurrentMethod, requireOutParamsAssigned: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: >DataFlowPass.Analyze(_compilation, methodSymbol, analyzedInitializers, diagsForCurrentMethod, requireOutParamsAssigned: false); [](start = 32, length = 127)

It feels like DiagnosticsPass.IssueDiagnostics should also be called here.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 2]
```diff
@@ -888,6 +888,14 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
 
                         if (body != null && ((methodSymbol.ContainingType.IsStructType() && !methodSymbol.IsImplicitConstructor) || instrumentForDynamicAnalysis))
                         {
+                            if (instrumentForDynamicAnalysis && methodSymbol.IsImplicitConstructor)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: >instrumentForDynamicAnalysis && methodSymbol.IsImplicitConstructor [](start = 32, length = 66)

It feels like the condition should be ```!methodSymbol.ContainingType.IsStructType() || methodSymbol.IsImplicitConstructor```.
In general, it looks like the logic here becomes rather hard to follow with code duplication on different levels of branching. Consider if we could separate the decision making about doing additional analysis and merging initializers into body into two independent ```if``` statements.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: The call to IssueDiagnostics on line 927 below will cover this.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: I see, thanks.

---
In reply to: [76301656](https://github.com/dotnet/roslyn/pull/13352#discussion_r76301656) [](ancestors = 76301656)

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: We can assert inside the if that the instrumentForDynamicAnalysis == true.

---
In reply to: [76295380](https://github.com/dotnet/roslyn/pull/13352#discussion_r76295380) [](ancestors = 76295380)

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: The condition is more correct as written. Were it to be changed as indicated above, then field initializers would get analyzed twice for explicit constructors of classes (once here, and once in FlowAnalysisPass.Rewrite).

I agree that the logic is hard to follow, but I'm not sure either that now is the best time to rearrange the code or that I am the most appropriate person to do it.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by tannergooding** [comment]
Signed build kicked off.
